### PR TITLE
New: Write CSV attachment from JSON action

### PR DESCRIPTION
**Pull request by kirill-levitskiy**
Add "Write CSV attachment from JSON Object" and "Write CSV attachment from JSON Array" actions
PR https://github.com/elasticio/csv-component/pull/56

### CODE REVIEW

File: component.json [hunk 1]
```diff
@@ -66,8 +66,47 @@
...
-            "properties": {}
+          "type": "object",
+          "properties": {}
+        },
+        "out": {}
+      }
+    },
+    "write_attachment_from_json": {
+      "description":
+      "Multiple incoming events can be combined into one CSV file with the write CSV action.  See https://github.com/elasticio/csv-component/ for additional documentation.",
```
Comment (jhorbulyk) [inline component.json:77]: There is the new help links section where links like this should be placed.

File: component.json [hunk 2]
```diff
@@ -66,11 +68,93 @@
...
+          "properties": {}
+        },
+        "out": {}
+      }
+    },
+    "write_attachment_from_json": {
+      "main": "./lib/actions/writeFromJson.js",
+      "title": "Write CSV attachment from JSON Object",
+      "help": {
+        "description": "Multiple incoming events can be combined into one CSV file with the write CSV action. See <a href='https://github.com/elasticio/csv-component/'>link</a> for additional documentation.",
```
Comment (jhorbulyk) [inline component.json:81]: https://docs.elastic.io/guides/configuring-help-links.html

File: lib/actions/writeFromJson.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      method: 'PUT',
+      timeout: REQUEST_TIMEOUT,
+      retry: REQUEST_MAX_RETRY,
+      delay: REQUEST_RETRY_DELAY,
+      maxContentLength: REQUEST_MAX_CONTENT_LENGTH,
+    });
+    stringifier.end();
+
+    const messageToEmit = messages.newMessageWithBody({
+      rowCount: finalRowCount,
```
Comment (jhorbulyk) [inline lib/actions/writeFromJson.js:96]: Can we add the steward URL to the body of the outgoing message?

File: lib/actions/writeFromArray.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    method: 'PUT',
+    timeout: REQUEST_TIMEOUT,
+    retry: REQUEST_MAX_RETRY,
+    delay: REQUEST_RETRY_DELAY,
+    maxContentLength: REQUEST_MAX_CONTENT_LENGTH,
+  });
+  stringifier.end();
+
+  const messageToEmit = messages.newMessageWithBody({
+    rowCount,
```
Comment (jhorbulyk) [inline lib/actions/writeFromArray.js:84]: Can we add the steward URL to the body of the outgoing message?

File: component.json (hunk 2 above)
Comment (kirill-levitskiy) [reply component.json:81]: Ok, got it. I'll remove second link

File: README.md [hunk 5]
```diff
@@ -100,6 +99,76 @@ The output of the CSV Write component will be a message with an attachment.  In
...
+### Write CSV attachment from Array
+
+* `Include Header` - this select configures output behavior of the component. If option is `Yes` or no value chosen than header of csv file will be written to attachment, this is default behavior. If value `No` selected than csv header will be omitted from attachment.
+* `Separator` - this select configures type of CSV delimiter in an output file. There are next options: `Comma (,)`, `Semicolon (;)`, `Space ( )`, `Tab (\t)`. 
+
+This action will convert an incoming array into a CSV file.
+
+The columns of the CSV file will be mapped with same names as keys in the input JSON body.
+The requirement for an input Array - each JSON object inside should be plain without nested complex objects and similar to each other. 
+The keys of an input JSON will be published as the header in the first row. For each incoming
```
[the remark this answers is not in the document]
Comment (kirill-levitskiy) [reply README.md:146]: Changed the description

File: README.md [hunk 6]
```diff
@@ -100,6 +99,76 @@ The output of the CSV Write component will be a message with an attachment.  In
...
+event, the value for each header will be `stringified` and written as the value
+for that cell. All other properties will be ignored. For example, headers
+`foo,bar` along with the following JSON events:
+
+```
+[
+{"foo":"myfoo", "bar":"mybar"},
+{"foo":"myfoo", "bar":[1,2]},
+{"bar":"mybar", "baz":"mybaz"}
```
[the remark this answers is not in the document]
Comment (kirill-levitskiy) [reply README.md:155]: This case will produce next csv:
```
foo
bar
```

File: lib/actions/writeFromArray.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
+const axios = require('axios');
+const csv = require('csv');
```
[the remark this answers is not in the document]
Comment (kirill-levitskiy) [reply lib/actions/writeFromArray.js:2]: I think a migration to other library should be implemented for all actions/triggers together. It is the idea for next task.

File: lib/actions/writeFromJson.js (hunk 3 above)
Comment (kirill-levitskiy) [reply lib/actions/writeFromJson.js:96]: Currently, the steward URL is in the `attachments` object. Sure, we can add it to the `body` object, but I think this all actions and triggers should have similar behavior, so we need to add this to all of them. This change can be implemented in next task.

File: lib/actions/writeFromArray.js (hunk 4 above)
Comment (kirill-levitskiy) [reply lib/actions/writeFromArray.js:84]: Currently, the steward URL is in the `attachments` object. Sure, we can add it to the `body` object, but I think this all actions and triggers should have similar behavior, so we need to add this to all of them. This change can be implemented in next task.

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,7 @@
+## 2.1.0 (April 5, 2020)
+
+* Add "Write CSV attachment from JSON" action
```
Comment (A3a3e1) [inline CHANGELOG.md:3]: * Update sailor version to 2.6.5

File: README.md [hunk 9]
```diff
@@ -100,6 +99,85 @@ The output of the CSV Write component will be a message with an attachment.  In
...
+* Payload will store data from Values of relevant Keys (Columns).
+
+Requirements:
+
+* The inbound message is an JSON Array of Object with identical structure;
+* Each JSON object has plain structure without nested levels. 
+
+The keys of an input JSON will be published as the header in the first row. For each incoming
+event, the value for each header will be `stringified` and written as the value
+for that cell. All other properties will be ignored. For example, headers
```
Comment (jhorbulyk) [inline README.md:157]: > All other properties will be ignored.

What are other properties?

Comment (kirill-levitskiy) [reply README.md:157]: Means, if the first object has keys: `Name`, `Surname` and the second - `Name`, `Surname`, `Email`, the `Email` property will be ignored and won't be added to an output file.
For example,
```
[
  {
    Name: 'John',
    Surname: 'Doe'
  },
  {
    Name: 'Jane',
    Surname: 'Doe',
    Email: 'jane.doe@email.domain'
  }
]
```
will be transformed into
```
Name,Surname
John,Doe
Jane,Doe
```

In opposite, if the first object has keys: `Name`, `Surname`, `Email` and the second - `Name`, `Surname`, all three properties will present in an output file.
```
[
  {
    Name: 'John',
    Surname: 'Doe',
    Email: 'john.doe@email.domain'
  },
  {
    Name: 'Jane',
    Surname: 'Doe'
  }
]
```
will be transformed into
```
Name,Surname,Emal
John,Doe,john.doe@email.domain
Jane,Doe,
```

**Pull request merged**